### PR TITLE
Support plugins extending Webpack build pipeline; Build translations from openproject-translations

### PR DESCRIPTION
**Pull request by myabc**
This PR concerns itself with two related, but separate themes:
### Support plugins extending Webpack build pipeline

https://community.openproject.org/work_packages/5697
https://community.openproject.org/work_packages/17241
#### Notes for review:
- Webpack will not bundle plugins (and fail silently) if a `Gemfile.lock` is inconsistent (for example, if there are conflict makers). I'm not how/if we should make this more obvious.
- Shelling out from node to Ruby is currently done **synchronously**, which is not really _the node way_, but I wasn't really sure how to handle callbacks in the webpack config.
### Build translations from openproject-translations, if available

https://community.openproject.org/work_packages/14679
#### Notes for review:
- ~~You can specify an `OPENPROJECT_TRANSLATIONS_ROOT` if you want to
  test this PR without adding openproject-translations as a dependency.~~
  
  _Removed in 4ea7cb0. Use a bundle override instead (e.g. `bundle config local.openproject-translations ~/git-repos/openproject-translations`_

---

See also 
- finnlabs/openproject-backlogs#114
- finnlabs/openproject-costs#115

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://feature-webpack-i18n-support.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

**Comment by myabc** [comment]
@ulferts @tessi How do we support translations for plugins that are neither German or English?

|  | core app | plugins |
| --- | --- | --- |
| bundled translations |  |  |
| en | X | X |
| de | X | X |
| openproject-translations |  |  |
| it | X | ? |
| es | X | ? |
| cn | X | ? |
| etc. | X | ? |

**Comment by ulferts** [comment]
@myabc don't know that, sorry.

**Comment by myabc** [comment]
Hound will continue to bark until #2184 is merged ... hint, hint... :poodle:

**Comment by tessi** [comment]
@myabc @ulferts I had a little chat with @kgalli about including plugin translations yesterday. He suggested to split the translation files from the plugin-code (and host the translations on some separate opf-related service). This would enable us to have many crowdin projects (and other translation services would probably work too), so that we can have one crowdin-project per plugin.

The translations-plugin could then fetch the translations of the current OpenProject version from that opf-service (which in turn asks crowdin) for the core and all plugins. In a second step, those translations could be combined into one `<locale>.yml` file.

You can probably ask @kgalli for details.

**Comment by myabc** [comment]
I've added notes to the description.

There are some additional, related tasks that I'll leave for other Pull Requests:
- bundling HTML templates along with plugin JS assets.
- moving Karma and Protractor tests under `frontend`

**Comment by myabc** [comment]
Additional, related tasks for translations that I'll leave for other Pull Requests:
- use  CrowdIn for German, English core translations.
- setting up CrowdIn projects for individual plugins.

File: bin/plugin_info [hunk 1]
```diff
@@ -26,25 +28,33 @@
...
+
+  Bundler.load.specs.each_with_object({}) do |spec, h|
+    if op_dep.include?(spec.name)
+      options = spec.source.options
+      h[spec.name] = spec.full_gem_path
+    end
+  end
+end
+
+puts JSON.generate(plugin_name_paths)
```
Comment (ulferts) [inline bin/plugin_info:60]: @myabc I don't think this is intended.

Comment (myabc) [inline bin/plugin_info:60]: @ulferts it is intended. The script should write to stdout. I guess `STDOUT.puts` might make the intention clearer.

Comment (ulferts) [inline bin/plugin_info:60]: @myabc what is intended to happen with that output? Why is it required to write to stdout? Why can't the loader of that file handle all of this internally?

Comment (myabc) [inline bin/plugin_info:60]: @ulferts `config/rails-plugins.conf.js` currently shells out to this script. If you can think a way of improving this, let me know.

Comment (ulferts) [inline bin/plugin_info:60]: I see... you have to get the output from ruby to js. No clean solution for that comes to mind. We could wrap it all in function that one can call with parameters. If no param is provided, nothing happens. The provided parameter specifies the means for transporting the information. But that would basically still be the same puts statement. Could be an ENV variable just the same but this does not make it better.

**Comment by ulferts** [comment]
looks good to me and works on my machine.

**Comment by 0xF013** [comment]
Looks ok to me

**Comment by myabc** [comment]
@ulferts issues you mentioned should be dealt with.

**Comment by ulferts** [comment]
Merging because I know that the failing cukes have nothing to do with this PR.

**Pull request merged**